### PR TITLE
Feature: test setup and books tests

### PR DESCRIPTION
**Pull request by AhsanDev7**
### **Problem Statement:**
There was a need to implement unit tests for the CRUD operations of the books model to ensure the reliability and functionality of the system.

### **Proposed Changes:**

- Set up unit testing for the project using pytest.
- Completed the factory setup for creating model instances in tests.
- Added unit tests to verify the functionality of the CRUD operations (Create, Read, Update, Delete) for the books, Authors, Publishers, and orders model.

### **Verification Steps:**

Run `pytest `to ensure all tests pass.

### CODE REVIEW

File: myapp/tests/factories.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            for author in extracted:
+                self.authors.add(author)
+
+    @factory.post_generation
+    def genres(self, create, extracted, **kwargs):
+        if not create:
+            return
+        if extracted:
+            for genre in extracted:
+                self.genres.add(genre)
```
Comment (iuxtl) [inline myapp/tests/factories.py]: Use [Traits](https://factoryboy.readthedocs.io/en/stable/reference.html#traits) for the generating side data.

File: myapp/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    books = BookFactory.create_batch(5, authors=[author], genres=[genre])
+    return books
+
+
+
+@pytest.fixture
+def user():
+    """Fixture to create a user for authentication."""
+    User = get_user_model()
+    return User.objects.create_user(username="testuser", password="password123")
```
Comment (yasirrose) [inline myapp/tests/test_views.py]: Create and use the user factory

File: requirements.txt [hunk 3]
```diff
@@ -5,10 +5,14 @@ djangorestframework-simplejwt==5.3.1
...
 django-filter==24.3
 django-debug-toolbar==4.4.6
+factory_boy==3.3.1
 gunicorn==23.0.0
 psycopg2-binary==2.9.10
 python-dateutil
 python-decouple==3.8
+pytest==8.3.4
+pytest-django==4.9.0
+pytest-factoryboy==2.7.0
```
Comment (iuxtl) [inline requirements.txt]: Why we need this package? `pytest-factoryboy==2.7.0`

File: myapp/tests/factories.py (hunk 1 above)
Comment (AhsanDev7) [reply myapp/tests/factories.py]: Implemented traits functionality in the factories.

File: myapp/tests/test_views.py (hunk 2 above)
Comment (AhsanDev7) [reply myapp/tests/test_views.py]: Created user factories and used it for user auth

File: requirements.txt (hunk 3 above)
Comment (AhsanDev7) [reply requirements.txt]: There wasn't a need for this package, I have uninstalled and removed it.

File: myapp/tests/factories.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
```
Comment (iuxtl) [inline myapp/tests/factories.py]: Tracts functionality was meant to be implemented for the side data, not for properties.

File: myapp/tests/test_books.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+@pytest.fixture
+def api_client():
+    return APIClient()
+
+
+@pytest.fixture
+def create_books():
+    author = AuthorFactory()
+    genre = GenreFactory()
+    return BookFactory.create_batch(5, authors=[author], genres=[genre])
```
Comment (iuxtl) [inline myapp/tests/test_books.py]: Use Traits to create side data

File: myapp/tests/test_publisher.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        assert response.status_code == 400
+        assert "name" in response.data
+        assert "established_year" in response.data
+        assert "website" in response.data
+        assert "contact_email" in response.data
+
+    def test_delete_publisher(self, auth_client):
+      """Test deleting a publisher, handling success and failure scenarios."""
+      publisher = PublisherFactory()
+      print(publisher)
```
Comment (iuxtl) [inline myapp/tests/test_publisher.py]: print statement left in code

File: myapp/tests/test_publisher.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+def auth_client():
+    """Fixture to authenticate the client using a user factory."""
+    user = UserFactory()
+    client = APIClient()
+    refresh = RefreshToken.for_user(user)
+    client.credentials(HTTP_AUTHORIZATION=f"Bearer {refresh.access_token}")
+    return client
+
+@pytest.mark.django_db
+class TestPublisherAPI:
```
Comment (iuxtl) [inline myapp/tests/test_publisher.py:28]: Where are the tests for the pagination?

File: myapp/tests/factories.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+fake = Faker()
+
+class UserFactory(DjangoModelFactory):
+    class Meta:
+        model = get_user_model()
+        django_get_or_create = ("username",)
+
+    username = factory.Faker("user_name")
+    email = factory.Faker("email")
+    password = factory.PostGenerationMethodCall("set_password", "password123")
```
Comment (iuxtl) [inline myapp/tests/factories.py]:
```suggestion
    from factory.django import Password
    password = Password("password")
    
```

File: myapp/tests/factories.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+            for author in extracted:
+                self.authors.add(author)
+
+    @factory.post_generation
+    def genres(self, create, extracted, **kwargs):
+        if not create:
+            return
+        if extracted:
+            for genre in extracted:
+                self.genres.add(genre)
```
Comment (iuxtl) [inline myapp/tests/factories.py]:
```suggestion
   class Params:
        with_profile = Trait(
            userprofile=RelatedFactory("core.tests.factories.UserProfileFactory", "user")
        )
```

**Pull request merged**